test/test_baggage.cpp

```diff
@@ -0,0 +1,214 @@
+            Baggage({{"key1", "value1"}, {"key2", "value2"}}),
+        },
+        {
+            "leading spaces 3",
+            "    key1    = value1,key2=value2",
+            Baggage({{"key1", "value1"}, {"key2", "value2"}}),
+        },
+        {
+            "leading spaces 4",
+            "    key1    = value1  ,key2=value2",
+            Baggage({{"key1", "value1"}, {"key2", "value2"}}),
+        },
+        {
+            "leading spaces 5",
+            "    key1    = value1  , key2=value2",
+            Baggage({{"key1", "value1"}, {"key2", "value2"}}),
+        },
+        {
+            "leading spaces 6",
+            "    key1    = value1  , key2  =value2",
+            Baggage({{"key1", "value1"}, {"key2", "value2"}}),
+        },
+        {
+            "leading spaces 7",
+            "    key1    = value1  , key2  =   value2",
+            Baggage({{"key1", "value1"}, {"key2", "value2"}}),
+        },
+        {
+            "leading spaces 8",
+            "    key1    = value1  , key2  =   value2  ",
+            Baggage({{"key1", "value1"}, {"key2", "value2"}}),
+        },
+        {
+            "leading spaces 9",
+            "key1   = value1,   key2=   value2",
+            Baggage({{"key1", "value1"}, {"key2", "value2"}}),
+        },
+        {
+            "spaces in key is allowed",
+            "key1 foo=value1",
+            Baggage({{"key1 foo", "value1"}}),
+        },
+        {
+            "verify separator",
+            "key1=value1;a=b,key2=value2",
+            Baggage({{"key1", "value1;a=b"}, {"key2", "value2"}}),
+        },
+        {
+            "malformed baggage",
+            ",k1=v1,k2=v2,",
+            Baggage::Error::MALFORMED_BAGGAGE_HEADER,
+        },
+        {
+            "malformed baggage 2",
+            "=",
+            Baggage::Error::MALFORMED_BAGGAGE_HEADER,
+        },
+        {
+            "malformed baggage 3",
+            "=,key2=value2",
+            Baggage::Error::MALFORMED_BAGGAGE_HEADER,
+        },
+        {
+            "malformed baggage 4",
+            "key1=value1,=",
+            Baggage::Error::MALFORMED_BAGGAGE_HEADER,
+        },
+
+    }));
+
+    CAPTURE(test_case.name, test_case.input);
+
+    const std::unordered_map<std::string, std::string> headers{
+        {"baggage", test_case.input}};
+    MockDictReader reader(headers);
+
+    auto maybe_baggage = Baggage::extract(reader);
+    if (maybe_baggage.has_value() && test_case.expected_baggage.has_value()) {
+      CHECK(*maybe_baggage == *test_case.expected_baggage);
+    } else if (maybe_baggage.if_error() &&
+               test_case.expected_baggage.if_error()) {
+      CHECK(maybe_baggage.error() == test_case.expected_baggage.error());
+    } else {
+      FAIL("mistmatch between what is expected and the result");
+    }
+  }
+}
+
+BAGGAGE_TEST("inject") {
+  SECTION("limit is respected") {
+    Baggage bag({{"violets", "blue"}, {"roses", "red"}});
```

Comment:
Can you also test cases like this where injection will properly URL encode characters? I suppose this also applies to parsing as well.

Here's a round-trip example: https://github.com/DataDog/system-tests/blob/68fb7b7f5adfd47e7c02eb3b34a86c6fe671f8e0/tests/parametric/test_headers_baggage.py#L104

Reply:
Replied in https://github.com/DataDog/dd-trace-cpp/pull/179/files#r1947301140